tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -0,0 +1,858 @@
+        // Seq(S)
+        val a = varPool.fresh
+        val sig = OperT1(Seq(SetT1(a)), SeqT1(a))
+        mkExRefApp(sig, Seq(arg))
+
+      case OperEx(op, args @ _*)
+          if op == TlaSetOper.in || op == TlaSetOper.notin =>
+        // x \in S, x \notin S
+        val a = varPool.fresh
+        val opsig = OperT1(List(a, SetT1(a)), BoolT1())
+        mkExRefApp(opsig, args)
+
+      case OperEx(op, args @ _*)
+          if op == TlaSetOper.subseteq || op == TlaSetOper.subsetProper
+            || op == TlaSetOper.supseteq || op == TlaSetOper.supsetProper =>
+        // S \subseteq T, S \subset T, S \supseteq T, S \supset T
+        val a = varPool.fresh
+        val opsig = OperT1(List(SetT1(a), SetT1(a)), BoolT1())
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaSetOper.SUBSET, args @ _*) =>
+        // SUBSET S
+        val a = varPool.fresh
+        val opsig = OperT1(List(SetT1(a)), SetT1(SetT1(a)))
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaSetOper.union, args @ _*) =>
+        // UNION S
+        val a = varPool.fresh
+        val opsig = OperT1(List(SetT1(SetT1(a))), SetT1(a))
+        mkExRefApp(opsig, args)
+
+      case OperEx(op, args @ _*)
+          if op == TlaSetOper.cap || op == TlaSetOper.cup || op == TlaSetOper.setminus =>
+        // S \\intersect T, S \\union T, S \\ T
+        val a = varPool.fresh
+        val opsig = OperT1(List(SetT1(a), SetT1(a)), SetT1(a))
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaSetOper.times, args @ _*) =>
+        // S \X T
+        val typeVars = varPool.fresh(args.length)
+        val opsig = OperT1(typeVars.map(SetT1(_)), SetT1(TupT1(typeVars: _*)))
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaSetOper.filter, bindingEx, bindingSet, pred) =>
+        // { x \in S: P }
+        // or, { <<x, ..., z>> \in S: P }
+        // the principal type is (a => Bool) => Set(a)
+        val bindings = translateBindings((bindingEx, bindingSet))
+        val typeVars = varPool.fresh(bindings.length)
+
+        // 1. When there is one argument, a set element has type "a", no tuple is involved.
+        // 2. When there are multiple arguments,
+        //    a set element has type type <<a, b>>, that is, it is a tuple
+        val funFrom =
+          if (bindings.length == 1) typeVars.head else TupT1(typeVars: _*)
+        // the principal type is ((a, b) => Bool) => Set(<<a, b>>) or just (a => Bool) => Set(a)
+        val principal = OperT1(Seq(OperT1(typeVars, BoolT1())), SetT1(funFrom))
+
+        // map implicitly introduces a lambda abstraction: λ x ∈ S, y ∈ T. P
+        val lambda = mkAbs(BlameRef(ex.ID), this(pred), bindings: _*)
+        // the resulting expression is ((((a, b) => Bool) => Set(<<a, b>>)) (λ x ∈ S, y ∈ T. P)
+        mkApp(ref, Seq(principal), lambda)
+
+      case OperEx(TlaSetOper.map, mapExpr, args @ _*) =>
+        // { x \in S, y \in T: e }
+        val (vars, sets) = args.zipWithIndex.partition(_._2 % 2 == 0)
```

Comment:
Shall we use `TlaOper.deinterleave` now?

Reply:
Good idea!